framework/src/base/MooseApp.C

```diff
@@ -347,37 +352,61 @@ MooseApp::MooseApp(InputParameters parameters)
 #ifdef HAVE_GPERFTOOLS
   if (isUltimateMaster())
   {
-    if (std::getenv("MOOSE_PROFILE_BASE") && std::getenv("MOOSE_HEAP_BASE"))
-      mooseError("Can not do CPU and heap profiling together");
-
+    static std::string profile_file;
     // For CPU profiling, users need to have envirement 'MOOSE_PROFILE_BASE'
     if (std::getenv("MOOSE_PROFILE_BASE"))
     {
-      static std::string profile_file =
-          std::getenv("MOOSE_PROFILE_BASE") + std::to_string(_comm->rank()) + ".prof";
-
       _cpu_profiling = true;
-
-      auto name = MooseUtils::splitFileName(profile_file);
-      if (!name.first.empty() && processor_id() == 0)
-        MooseUtils::makedirs(name.first.c_str());
-
-      if (!ProfilerStart(profile_file.c_str()))
-        mooseError("CPU profiler is not started properly");
+      profile_file = std::getenv("MOOSE_PROFILE_BASE") + std::to_string(_comm->rank()) + ".prof";
     }
 
     // For Heap profiling, users need to have 'MOOSE_HEAP_BASE'
     if (std::getenv("MOOSE_HEAP_BASE"))
     {
-      static std::string profile_file =
-          std::getenv("MOOSE_HEAP_BASE") + std::to_string(_comm->rank());
-
       _heap_profiling = true;
+      profile_file = std::getenv("MOOSE_HEAP_BASE") + std::to_string(_comm->rank()) + ".prof";
+    }
+
+    if (_cpu_profiling && _heap_profiling)
+      mooseError("Can not do CPU and heap profiling together");
 
+    if (_cpu_profiling || _heap_profiling)
+    {
       auto name = MooseUtils::splitFileName(profile_file);
       if (!name.first.empty() && processor_id() == 0)
         MooseUtils::makedirs(name.first.c_str());
+      _comm->barrier();
+    }
+
+    if (isParamValid("gperf_profiler_on"))
+    {
+      auto rankstr = getParam<std::string>("gperf_profiler_on");
+      std::vector<processor_id_type> ranks;
+      bool success = MooseUtils::tokenizeAndConvert(rankstr, ranks, ", ");
+      if (!success)
+        mooseError("Invalid argument for --gperf-profiler-on: '", rankstr, "'");
+      for (auto & rank : ranks)
+      {
+        if (rank >= _comm->size())
+          mooseError("Invalid argument for --gperf-profiler-on: ",
+                     rank,
+                     " is greater than or equal to ",
+                     _comm->size());
+        _gperf_rank_set.insert(rank);
+      }
+      if (_gperf_rank_set.size() != ranks.size())
+        mooseWarning("Ranks for --gperf-profiler-on are specified multiple times");
+    }
+    else
+      for (processor_id_type i = 0; i < _comm->size(); ++i)
+        _gperf_rank_set.insert(i);
```

Comment:
```
      for (processor_id_type i = 0; i < _comm->size(); ++i)
         _gperf_rank_set.insert(i)
```

should change to 

`_gperf_rank_set.insert(_comm->rank())`

There is no reason to add the whole world into `_gperf_rank_set`

Reply:
ok, I guess we can remove `_gperf_rank_set` because we will not use it later.